8226236: [TESTBUG] win32: gc/metaspace/TestCapacityUntilGCWrapAround.java fails

YaSuenag opened the pull request:
Originally filed at AdoptOpenJDK:
https://github.com/AdoptOpenJDK/openjdk-tests/issues/1162

The test fails on 32bit windows with:

```
java.lang.IllegalStateException: WB_IncMetaspaceCapacityUntilGC: could not increase capacity until GC due to contention with another thread
        at sun.hotspot.WhiteBox.incMetaspaceCapacityUntilGC(Native Method)
        at TestCapacityUntilGCWrapAround.main(TestCapacityUntilGCWrapAround.java:51)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
        at java.lang.Thread.run(Thread.java:748)
```

`TestCapacityUntilGCWrapAround` passes `4GB - 1` to `incMetaspaceCapacityUntilGC()`. It seems to be too big.  
And also this code seems to want to check the behavior when `_capacity_until_gc` is overflown. White box test would throw ISE when it hapen. So we need to handle it correctly.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (8/8 passed) |

### Issue
 * [JDK-8226236](https://bugs.openjdk.java.net/browse/JDK-8226236): [TESTBUG] win32: gc/metaspace/TestCapacityUntilGCWrapAround.java fails


### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/628/head:pull/628`
`$ git checkout pull/628`

shipilev reviewed (commented):
Current patch makes `gc/metaspace` tests pass on `x86_32` for me. They used to fail as described.

YaSuenag commented:
/label hotspot-gc

YaSuenag commented:
Thanks @shipilev ! Can you approve this change?

shipilev commented:
I'd prefer @tstuefe and @tschatzl to have a look.

tstuefe commented on test/hotspot/jtreg/gc/metaspace/TestCapacityUntilGCWrapAround.java:
```diff
@@ -41,19 +40,30 @@
         long before = wb.metaspaceCapacityUntilGC();
-        // Now force possible overflow of capacity_until_GC.
-        long after = wb.incMetaspaceCapacityUntilGC(MAX_UINT);
+        long after = wb.incMetaspaceCapacityUntilGC(100 * MB);
```
The test makes only sense with 4G increment. AFAIU the counter is supposed to overflow, and cap at MaxMetaspaceSize instead.

tstuefe reviewed (changes requested):
Hi Yasumasa,

sorry, I do not understand what the error is yet and to me the patch does seem wrong in a number of places. 

Both test and whitebox function seem to be doing what they are supposed to do.

Is this error intermittent or always reproducable? If the latter, since when (since the code seems old)? 

Have the AdoptOpenJDK folks done some pre-analysis? 

----

From just looking at the code (cannot build 32bit right now), WB_IncMetaspaceCapacityUntilGC() seems correct to me.

- in comes a jlong with 4G-1, so FFFF-FFFF
- it fits into a 32bit size_t so its fine
- we calculate aligned_inc by aligning the value down somewhat, probably a page: FFFF-F000
- we feed this value FFFF-F000 as *inc* value to  MetaspaceGC::inc_capacity_until_GC
- it returns false. We throw an exception.

In MetaspaceGC::inc_capacity_until_GC:

```
bool MetaspaceGC::inc_capacity_until_GC(size_t v, size_t* new_cap_until_GC, size_t* old_cap_until_GC, bool* can_retry) {
  assert_is_aligned(v, Metaspace::commit_alignment());

  size_t old_capacity_until_GC = _capacity_until_GC;
  size_t new_value = old_capacity_until_GC + v;

  if (new_value < old_capacity_until_GC) {
    // The addition wrapped around, set new_value to aligned max value.
    new_value = align_down(max_uintx, Metaspace::commit_alignment());
  }

  if (new_value > MaxMetaspaceSize) {
    if (can_retry != NULL) {
      *can_retry = false;
    }
    return false;
  }
```

we enter with size_t v = FFFF-F000

We calc the new value. Since new value = old value + inc, this will overflow

We sense the overflow and correct the new value to be max_uintx aligned down by commit alignment.

Which will be smaller than MaxMetaspaceSize. Test does not set it therefore it is max_uintx.

So far all good IIUC.


```
  if (can_retry != NULL) {
    *can_retry = true;
  }
  size_t prev_value = Atomic::cmpxchg(&_capacity_until_GC, old_capacity_until_GC, new_value);

  if (old_capacity_until_GC != prev_value) {
    return false;
  }

  if (new_cap_until_GC != NULL) {
    *new_cap_until_GC = new_value;
  }
  if (old_cap_until_GC != NULL) {
    *old_cap_until_GC = old_capacity_until_GC;
  }
  return true;
```

From looking at this code, the only way I can see this function returning false would be if a concurrent thread modified this threshold. Which should be super rare. 

Hence my initial question about intermittentness. If its reproducable I do not think I understand it yet.

Thanks, Thomas

YaSuenag commented:
I think TestCapacityUntilGCWrapAround.java has two problems:

1. It might attempt to increase metaspace size over MaxMetaspaceSize (this PR fixes it)
2. Overflow test would always fail because `MetaspaceGC::inc_capacity_until_GC()` always returns `false`

This test has introduced in JDK-8049599, but I cannot know details because JBS ticket is closed, but we can see its commit: https://github.com/openjdk/jdk/commit/6f4355a3a6be2482c388c2022fc7c3c4b2f481c5  
I enabled this test on 64bit platforms because these problems might happen on them.

If you agree with the fix for 1. , I will add the test for 2. to this PR.

tstuefe commented:
I found @shipilev's analysis in the bug report and understand better what happens now. 

I do not think this is a test error, the test is correct and shows a real issue. Just pasting my JBS comment here:

The test verifies that calling MetaspaceGC::inc_capacity_until_GC() repeatedly will not overflow the gc threshold.

1) MetaspaceGC::ergo_initialize()

MaxMetaspaceSize, if left unspecified, is supposed to be "infinite". In reality it defaults to max_uintx. But it gets aligned down by Metaspace::reserve_alignment. That one is either one of os::vm_allocation_granularity or 4 pages, whatever is larger. The 4 pages thing I recently added with JDK-8245707 to shake loose misuse of this alignment value (exactly cases like this).

So on Windows, Metaspace::reserve_alignment has always been os::vm_allocation_granularity (64K). On Linux, it was always 4K, until recently it switched to 16K.

This explains the 16K aligned value we see for MaxMetaspaceSize in Alexeys test: 4294950912 (FFFFC000).

2) MetaspaceGC::inc_capacity_until_GC()

The increase value causes an overflow - its supposed to do that. Overflow gets handled by:

  if (new_value < old_capacity_until_GC) {
    // The addition wrapped around, set new_value to aligned max value.
    new_value = align_down(max_uintx, Metaspace::commit_alignment());
  }

which sets the new threshold at max_uintx aligned down by Metaspace::commit_alignment(), which is just os::vm_page size.

So the new value is 4294963200, resp. FFFFF000

So the problem is that one value gets aligned down by Metaspace::reserve_alignment() (os::vm_allocation_granularity()), the other by Metaspace::commit_alignment (os::vm_page_size()). Then we compare those values.

So the test is okay. It showed us a real issue. I am not 100% sure what the correct behavior would be. Maybe increasing the gc threshold beyond MaxMetaspaceSize should not be an error at all, but we should just cap out at that value?

Possible solutions:

A) As stated above, just cap.

b) All that alignment business is actually unnecessary and we could remove it for clarity after JEP387 is out of the door. MaxMetaspaceSize does not need to be aligned to anything, neither does the GC threshold.

C) as a small workaround, at (2) one probably could use Metaspace::reserve_alignment, not commit_alignment. That makes no sense semantically but would remove this error.

(This must have been always a problem on Windows, and only recently on Linux, right?)

tstuefe commented:
A minimal fix - which preserves the current behavior as much as possible:

--- a/src/hotspot/share/memory/metaspace.cpp
+++ b/src/hotspot/share/memory/metaspace.cpp
@@ -152,7 +152,7 @@ bool MetaspaceGC::inc_capacity_until_GC(size_t v, size_t* new_cap_until_GC, size
 
   if (new_value < old_capacity_until_GC) {
     // The addition wrapped around, set new_value to aligned max value.
-    new_value = align_down(max_uintx, Metaspace::commit_alignment());
+    new_value = align_down(max_uintx, Metaspace::reserve_alignment());
   }
 
   if (new_value > MaxMetaspaceSize) {

This preserves the current behavior:
- increasing the gc threshold beyond MaxMetaspaceSize should be an error
- on 32bit, increasing the gc threshold beyond the scope of the 32bit counter should be tolerated and result in a value capped at the end of 32bit; unless MaxMetaspaceSize is set and lower than max, in which case it would be an error.

Note that I have no possibility to test this since the 32bit build on Windows still seems broken.

YaSuenag commented:
Thanks @tstuefe for the fix! I will merge it to this PR.

> (This must have been always a problem on Windows, and only recently on Linux, right?)

I don't know how long this problem has been available, but I saw this on Linux x64 when I removed `vm.bits == 32`. It is caused by new_value exceeds MaxMetaspaceSize.

I think we should be current behavior should be preserved. I added the logic which relates to `_capacity_until_gc`  in [JDK-8217432](https://bugs.openjdk.java.net/browse/JDK-8217432). Before this fix, I saw OutOfMemoryError despite that the memory was still available. And also @shipilev seems to had seen [the problems with Metaspace on Epsilon](https://mail.openjdk.java.net/pipermail/hotspot-gc-dev/2019-January/024595.html).

BTW TestCapacityUntilGCWrapAround.java still should be run on 32bit platforms only? I think we should check whether `_capacity_until_gc`  exceeds MaxMetaspaceSize on 64bit platforms.

tstuefe commented:
Hi Yasumasa,

> Thanks @tstuefe for the fix! I will merge it to this PR.
> 

No I think this is the fix. I still think the test is correct and does what it is supposed to do, see below.

> > (This must have been always a problem on Windows, and only recently on Linux, right?)
> 
> I don't know how long this problem has been available, but I saw this on Linux x64 when I removed `vm.bits == 32`. It is caused by new_value exceeds MaxMetaspaceSize.

You see this on 64bit because the test, as it is coded now, is not meant to run on 64bit.

When running on 64bit first you will trigger this exception:
```
  jlong max_size_t = (jlong) ((size_t) -1);
  if (inc > max_size_t) {
    THROW_MSG_0(vmSymbols::java_lang_IllegalArgumentException(),
        err_msg("WB_IncMetaspaceCapacityUntilGC: inc does not fit in size_t: " JLONG_FORMAT, inc));
  }
```
because it was written with 32bit in mind. Its purpose is to check if the inc value - itself a jlong - exceeds the range for a 32bit unsigned. It uses a jlong - 64bit signed - as comparison variable and sets it to (size_t)-1, clearly to get a 32bit SIZE_MAX. On 64bit it is oc a real -1, so the comparison will always be true.

Once you fix that or comment it out, on 64bit you will trigger this test assertion:
```
        Asserts.assertLTE(after, MAX_UINT,
                          "Increasing with MAX_UINT should not cause value larger than MAX_UINT:" + after);
```
because we just increased the gc threshold beyond the 32bit MAX_UINT value coded in the test.

Granted, one could make this test cleaner and make it work on 64bit. But what for? This overflow is only an issue on 32bit. It theoretically could be an issue for 64bit too, if you call the increase function just very very often enough, but I think that is a bit far fetched.

If you enable the test for 64 bit you have to make sure that you call incMetaspaceCapacityUntilGC() with either a high enough increase or often enough to ensure that you hit the limit of a 64bit gc threshold counter. And then fix the overflow handling. But I seriously think this is not a real world issue.

Alternatively one could #ifdef the whole whitebox function out for 64bit, just compile it on 32bit. 

> 
> I think we should be current behavior should be preserved. I added the logic which relates to `_capacity_until_gc` in [JDK-8217432](https://bugs.openjdk.java.net/browse/JDK-8217432). Before this fix, I saw OutOfMemoryError despite that the memory was still available. And also @shipilev seems to had seen [the problems with Metaspace on Epsilon](https://mail.openjdk.java.net/pipermail/hotspot-gc-dev/2019-January/024595.html).
> 
> BTW TestCapacityUntilGCWrapAround.java still should be run on 32bit platforms only? I think we should check whether `_capacity_until_gc` exceeds MaxMetaspaceSize on 64bit platforms.

Look at it this way:

The tests are called with MaxMetaspaceSize unset, hence "infinite". Therefore the VM should *never* run into a situation where it thinks that the GC threshold is larger than MaxMetaspaceSize. Regardless what input values we feed into the increase function. The test feeds it very large values on purpose. Reducing the inc values defeats the purpose of this test.

The error is caused because, technically, MaxMetaspaceSize="infinite" is in reality implemented by MaxMetaspaceSize="very large value but a bit smaller than SIZE_MAX because of alignment". Which gives a small window, in this case of 16K, for values to be larger than MaxMetaspaceSize.

And then the overflow handling is broken and causes, as part of its value correction, the gc threshold to fall into this 16K window and be larger than MaxMetaspaceSize. And that exactly is the error.

The test really could be clearer and have comments and such, and the GC metaspace coding too. But the test is not wrong.

YaSuenag commented:
Ok, I will remove the change for 64bit platforms. (I aimed to check the case that MaxMetaspaceSize was specified, but it seems to be unneeded.)  
And thanks for explanation! I understood why `new_value` exceeds MaxMetaspaceSize in JBS case. I will push new change.

@shipilev Could you check again with this change?

shipilev commented:
> @shipilev Could you check again with this change?

Yes, it passes `gc/metaspace` tests on both `x86_32` and `x86_64`.

tstuefe reviewed (approved):
Thanks for your patience, Yasumasa. LGTM.

tstuefe commented:
(Created https://bugs.openjdk.java.net/browse/JDK-8254765 to track a rework of this gc threshold handling)

YaSuenag commented:
/integrate
